modules/nextflow/src/main/groovy/nextflow/script/WorkflowDef.groovy

```diff
@@ -234,23 +253,179 @@ class WorkflowParamsResolver {
+        _into_topic(out[0], name)
     }
+
+}
+
+/**
+ * Implements the DSL for publishing workflow outputs
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+@CompileStatic
+class WorkflowPublishDsl {
+
+    private static final List<String> PUBLISH_OPTIONS = List.of(
+        'contentType',
+        'enabled',
+        'ignoreErrors',
+        'mode',
+        'overwrite',
+        'pattern',
+        'storageClass',
+        'tags'
+    )
+
+    private Binding binding
+
+    private Path directory = Paths.get('.').complete()
+
+    private Map defaults = [:]
+
+    private boolean directoryOnce = false
+
+    WorkflowPublishDsl(Binding binding) {
+        this.binding = binding
+    }
+
+    @Override
+    Object getProperty(String name) {
+        try {
+            return binding.getProperty(name)
+        }
+        catch( MissingPropertyException e ){
+            return super.getProperty(name)
+        }
+    }
+
+    void directory(Map defaults=[:], String directory) {
+        if( directoryOnce )
+            throw new ScriptRuntimeException("Output directory cannot be defined more than once in the workflow output definition")
+        directoryOnce = true
+
+        this.directory = (directory as Path).complete()
+        this.defaults = defaults
+    }
+
+    void path(String path, Closure closure) {
+        final dsl = new PathDsl(directory.resolve(path), defaults)
+        final cl = (Closure)closure.clone()
+        cl.setResolveStrategy(Closure.DELEGATE_FIRST)
+        cl.setDelegate(dsl)
+        cl.call()
+    }
+
+    class PathDsl {
+
+        private Path path
+        private Map defaults
+        private boolean defaultsOnce = false
+
+        PathDsl(Path path, Map defaults) {
+            this.path = path
+            this.defaults = defaults
+        }
+
+        void defaults(Map opts) {
+            if( defaultsOnce )
+                throw new ScriptRuntimeException("Publish defaults cannot be defined more than once for a given path")
+            defaultsOnce = true
+
+            validatePublishOptions(opts)
+            defaults.putAll(opts)
+        }
+
+        void path(String subpath, Closure closure) {
+            final dsl = new PathDsl(path.resolve(subpath), defaults)
+            final cl = (Closure)closure.clone()
+            cl.setResolveStrategy(Closure.DELEGATE_FIRST)
+            cl.setDelegate(dsl)
+            cl.call()
+        }
+
+        void from(Map opts=[:], DataflowWriteChannel source) {
+            validatePublishOptions(opts)
+            if( opts.ignoreErrors )
+                opts.failOnError = !opts.remove('ignoreErrors')
+            new PublishOp(CH.getReadChannel(source), defaults + opts + [path: path]).apply()
+        }
+
+        void from(Map opts=[:], ChannelOut out) {
+            if( out.size() != 1 )
+                throw new IllegalArgumentException("Cannot publish a multi-channel output")
+            from(opts, out[0])
+        }
```

Comment:
@pditommaso what do you think about allowing a multi-channel output to be published

e.g. if I specify `from foo.out` and `foo` has multiple outputs, then publish all of them under the hood

Currently it is only allowed if the process has only one output

Reply:
or, prompt to be explicit on the channel of interest:

`Cannot publish a multi-channel output` -> `This is a multi-channel output: please specify the channel you want to publish.`

I would lean towards one or the other option also based on typical nf-core usage: do multiple channels from a process get typically published in the exact way (same output path, same options..)?  If so, I would prefer Ben's.   
Otherwise, if output statements typically need to be distinct, probably better to prompt the pipeline dev to be specific.

Reply:
I did encounter some cases in rnaseq where an entire process was published with no extra options. In this case it would be nice to just say `from foo.out` instead of enumerating all of the outputs of `foo`

Reply:
Yes, it should be specified the punctual one, either with array notation or emit name

Reply:
Fine, It will just be more verbose. But probably better in this case to be explicit and robust to e.g. adding a process output that you don't intend to publish